op-node/rollup/finality/finalizer.go

```diff
@@ -300,6 +287,98 @@ func (fi *Finalizer) tryFinalize() {
 	}
 }
 
+/*
+ *  tryFinalizeOnConsecutiveBlockQuorom tries to finalize L2 blocks in consecutive order.
+ *
+ *  If there are any L2 blocks that are not finalized, then we can't finalize the later ones.
+ *  This is because we need to finalize the L2 blocks in order to guarantee consecutive quorom.
+ *
+ *  It queries the Babylon gadget to check if the L2 blocks are finalized.
+ *  If all the L2 blocks in the range are finalized, then it sets the last finalized
+ *  L2 block and the derived-from L1 block.
+ *
+ *  If some of the L2 blocks in the range are not finalized, then it stops iterating.
+ *
+ *  If the error encountered is of type NoFpHasVotingPowerError, it should be ignored;
+ *  for any other error types, emit an critical error event.
+ */
+func tryFinalizeOnConsecutiveBlockQuorom(
+	fd FinalityData,
+	fi *Finalizer,
+	finalizedL2 *eth.L2BlockRef,
+	finalizedDerivedFrom *eth.BlockID,
+) (bool, error) {
+	blockCount := int(fd.L2Block.Number - finalizedL2.Number)
+	l2Blocks := make(map[uint64]eth.L2BlockRef)
+	queryBlocks := make([]*cwclient.L2Block, blockCount)
+
+	for i := 0; i < blockCount; i++ {
+		blockNumber := uint64(i) + finalizedL2.Number + uint64(1)
+		l2Block, err := fi.l2Fetcher.L2BlockRefByNumber(fi.ctx, blockNumber)
+		if err != nil {
+			fi.emitter.Emit(rollup.CriticalErrorEvent{Err: fmt.Errorf(
+				"failed to check if block %d to %d is finalized on Babylon, could not fetch block %d: %w",
+				finalizedL2.Number+1,
+				fd.L2Block.Number,
+				blockNumber,
+				err,
+			)})
+			return false, err
+		}
+		l2Blocks[blockNumber] = l2Block
+
+		queryBlocks[i] = &cwclient.L2Block{
+			BlockHeight:    l2Block.Number,
+			BlockHash:      l2Block.Hash.String(),
+			BlockTimestamp: l2Block.Time,
+		}
+		fi.log.Debug(
+			"added block to babylon gadget's query params",
+			"block_height", queryBlocks[i].BlockHeight,
+			"block_hash", queryBlocks[i].BlockHash,
+			"block_timestamp", queryBlocks[i].BlockTimestamp,
+		)
+	}
+
+	lastFinalizedBlockNumber, err := fi.babylonFinalityClient.QueryBlockRangeBabylonFinalized(queryBlocks)
+
+	// TODO: we shouldn't skip on no voting power.
+	// https://github.com/babylonchain/babylon-finality-gadget/issues/59
+	if err != nil && !errors.Is(err, sdkclient.ErrNoFpHasVotingPower) {
```

Comment:
This seems not consistent with the logic in Babylon. Babylon considers a block not finalisable if no FP has voting power at this height ([ref](https://github.com/babylonchain/babylon/blob/fe247dd38a83ccbf940dabed5dafaa69698b9eb0/x/finality/keeper/tallying.go#L62-L65)). This is because no BTC can be slashed at this height anyway. 

But maybe we need to have a different strategy for L2s. It's sequencer's responsibility to keep having BTC stake such that users will trutst him. Once sequencer does not have any BTC stake backed then users might not trust him anyway, and the L2 might need to stall. Not entirely sure about this part.

Maybe for now we can skip to limit the impact of BTC staking on L2's liveness?

Reply:
@SebastianElvis yes you are right. I realized this problem yesterday and created https://github.com/babylonchain/babylon-finality-gadget/issues/59

Reply:
> Maybe for now we can skip to limit the impact of BTC staking on L2's liveness?

I'd vote not to skip for security concern I described in https://github.com/babylonchain/babylon-finality-gadget/pull/38#issuecomment-2231526118

the original purpose of adding this bypass is to consider cases when the BTC staking is not activated. but I missed the fact that voting power can get to zero again if all FPs got slashes.

so the right fix is to query BTC staking activation height instead of relying on voting power

Reply:
after https://github.com/babylonchain/babylon-finality-gadget/issues/59, the chain will stall on any error querying the SDK. not just `ErrNoFpHasVotingPower`. added the comments in https://github.com/babylonchain/optimism/commit/20ce4dce5061ca0a0da70f6868c66bb939f9570d